server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java

```diff
@@ -186,7 +206,12 @@ public void close() {
 
             @Override
             public boolean isDone() {
-                return chunkIterator.hasNext() == false;
+                var result = chunkIterator.hasNext() == false;
+                if (result && sizeConsumer != null) {
+                    sizeConsumer.accept(size);
+                    sizeConsumer = null;
+                }
```

Comment:
I'd rather we added some logic to `DefaultRestChannel` to grab the size when the response is closed - for instance we may not get all the way to `isDone()` before the client closes the channel.

Reply:
Thanks for the prompt. I have now moved these logic entirely into RestController and I think it looks better than before. I considered DefaultRestChannel as well, but I think having it done within RestController is better due to easier access to different pieces of information.